pageserver/src/tenant/blob_io.rs

```diff
@@ -211,6 +237,62 @@ impl<const BUFFERED: bool> BlobWriter<BUFFERED> {
         (src_buf, Ok(()))
     }
 
+    pub async fn write_compressed_blob(
+        &mut self,
+        srcbuf: Bytes,
+        compress: bool,
+    ) -> Result<u64, Error> {
+        let offset = self.offset;
+
+        let len = srcbuf.len();
+
+        let mut io_buf = self.io_buf.take().expect("we always put it back below");
+        io_buf.clear();
+        let mut is_compressed = false;
+        if len < 128 {
+            // Short blob. Write a 1-byte length header
+            io_buf.put_u8(NO_COMPRESSION);
+            io_buf.put_u8(len as u8);
+        } else {
+            // Write a 4-byte length header
+            if len > 0x7fff_ffff {
+                return Err(Error::new(
+                    ErrorKind::Other,
+                    format!("blob too large ({} bytes)", len),
+                ));
+            }
+            if compress && len == BLCKSZ as usize {
+                let compressed = lz4_flex::block::compress(&srcbuf);
+                if compressed.len() < len {
```

Comment:
For large volumes data that's already compressed (i.e. is incompressible for us), it is expensive to compress each page and then decide whether to throw the result away.  It would be more efficient to have some state in the layer writer that switches off compression after the first few pages in an image layer, if the compression ratio achieved is not worthwhile.

The threshold for "not worthwhile" is probably not as simple as "compressed.len < len" -- consider a page that contains rows where the  user data is compressed, but has some postgres envelope content that is compressible: we might get e.g. 0.95 compression ratio -- in this case it is probably not worth the CPU cost to compress the data.

This state should reset when we cross a relation boundary, as the compressibilty of data from one relation is independent of the compressibility of data in another.

Reply:
IMHO it sounds to be too complicated and non-reliable.  Cost of compression is not so high comparing with all other manipulations with page images. So situation when we try to compress page and that store original page is not something very awful. And I expect that it will rarely happen, because even even TOAST pages compressed using internal Postgres lz compressor can be a little bit compressed so that it's size become smaller than original page.

And decompression speed is very high, so even instead of 8192 bytes we store compressed image with with size 7999 bytes, still it will not affect speed.